mutters-templated-intent/src/main/java/com/rabidgremlin/mutters/templated/TemplatedUtterance.java

```diff
@@ -1,127 +1,178 @@
-package com.rabidgremlin.mutters.templated;
-
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
-
-import org.apache.commons.lang.StringUtils;
-
-import com.rabidgremlin.mutters.core.Context;
-import com.rabidgremlin.mutters.core.Slot;
-import com.rabidgremlin.mutters.core.SlotMatch;
-import com.rabidgremlin.mutters.core.Slots;
-
-/**
- * This class handles the matching of an utterance against an utterance template.
- * 
- * It also handles the extraction slot data.
- * 
- * @author rabidgremlin
- *
- */
-public class TemplatedUtterance
-{
-  private String template;
-
-  private HashSet<String> slotNames = new HashSet<String>();
-
-  private Pattern matchPattern;
-
-  public TemplatedUtterance(String[] tokenizedTemplate)
-  {
-    this.template = StringUtils.join(tokenizedTemplate, ' ');
-
-    String regexStr = "^";
-
-    for (String token : tokenizedTemplate)
-    {
-      if (token.startsWith("{") && token.endsWith("}"))
-      {
-        String slotName = token.substring(1, token.length() - 1);
-
-        regexStr += "(?<" + slotName + ">.*) ";

```

Comment:
looks like it was doing string concat in loop before (could explain why it was slow) did you run a benchmark tool?